jobmaster/cvsJob/cvsJobMaster.go

```diff
@@ -82,87 +92,132 @@ func NewCVSJobMaster(ctx *dcontext.Context, workerID lib.WorkerID, masterID lib.
+		status := handle.Status()
 		switch status.Code {
 		case lib.WorkerStatusNormal, lib.WorkerStatusFinished, lib.WorkerStatusStopped:
-			num, err := strconv.ParseInt(string(status.ExtBytes), 10, 64)
+			taskStatus := &cvsTask.Status{}
+			err := json.Unmarshal(status.ExtBytes, taskStatus)
 			if err != nil {
 				return err
 			}
-			worker.curLoc = num
-			jm.counter += num
-			log.L().Debug("cvs job tmp num ", zap.Any("id", worker.handle.ID()), zap.Int64("counter", num), zap.Any("status", status.Code))
-			// todo : store the sync progress into the meta store for each file
-			if status.Code == lib.WorkerStatusFinished {
-				filesNum++
-			}
+
+			jm.jobStatus.FileInfos[idx].Location = taskStatus.CurrentLoc
+			jm.counter += taskStatus.Count
+
+			log.L().Debug("cvs job tmp num ", zap.Any("id", idx), zap.Any("status", string(status.ExtBytes)))
 		case lib.WorkerStatusError:
-			log.L().Error("sync file failed ", zap.Any("idx", worker.idx))
+			log.L().Error("sync file failed ", zap.Any("idx", idx))
 		default:
 			log.L().Info("worker status abnormal", zap.Any("status", status))
 		}
 	}
 	if jm.statusRateLimiter.Allow() {
+		statsBytes, err := json.Marshal(jm.jobStatus)
+		if err != nil {
+			log.L().Warn("serialize job status failed, try next time", zap.Any("master id", jm.workerID), zap.Error(err))
+			return err
+		}
+		_, err = jm.MetaKVClient().Put(ctx, jm.workerID, string(statsBytes))
+		if err != nil {
+			log.L().Warn("update job status failed, try next time", zap.Any("master id", jm.workerID), zap.Error(err))
+		}
 		log.L().Info("cvs job master status", zap.Any("id", jm.workerID), zap.Int64("counter", jm.counter), zap.Any("status", jm.getStatusCode()))
 	}
 	if jm.getStatusCode() == lib.WorkerStatusStopped {
 		log.L().Info("cvs job master stopped")
 		return jm.BaseJobMaster.Exit(ctx, jm.Status(), nil)
 	}
-	if filesNum == jm.filesNum {
-		jm.setStatusCode(lib.WorkerStatusFinished)
-		log.L().Info("cvs job master finished")
-		return jm.BaseJobMaster.Exit(ctx, jm.Status(), nil)
-	}
 	return nil
 }
 
-func (jm *JobMaster) OnMasterRecovered(ctx context.Context) error {
+func (jm *JobMaster) OnMasterRecovered(ctx context.Context) (err error) {
+	log.L().Info("recovering job master", zap.Any("id", jm.ID()))
+	// load self status
+	resp, err := jm.MetaKVClient().Get(ctx, jm.workerID)
+	if err != nil {
+		log.L().Warn("load status failed", zap.Any("master id", jm.ID), zap.Error(err))
+		return err
+	}
+	if len(resp.Kvs) != 1 {
+		log.L().Error("jobmaster meta unexpected result", zap.Any("master id", jm.ID()), zap.Any("meta counts", len(resp.Kvs)))
+	}
+	statusBytes := resp.Kvs[0].Value
+	log.L().Info("jobmaster recover from meta", zap.Any("master id", jm.ID()), zap.String("status", string(statusBytes)))
+	err = json.Unmarshal(statusBytes, jm.jobStatus)
+	if err != nil {
+		return err
+	}
+	for id := range jm.jobStatus.FileInfos {
+		info := &WorkerInfo{}
+		info.needCreate.Store(true)
```

Comment:
If the workers are still running, will the job master create duplicated workers?

Reply:
No, when "jm.IsMasterReady()" match, it implies all the onlined workers have been notified to job master, and others are sure to be dead.
When worker is online, it will set "needCreate" to false.